docs/source/blueprints/BP002.rst

```diff
@@ -0,0 +1,220 @@
+BP002
+#####
+
+:Number: BP002
+:Title: Requirements resolver
+:Author: Willian Rampazzo <willianr@redhat.com>
+:Discussions-To: https://github.com/avocado-framework/avocado/issues/3455
+:Reviewers:
+:Created: 27-Jan-2020
+:Type: Architecture Blueprint
+:Status: Draft
+
+.. contents:: Table of Contents
+
+TL;DR
+*****
+
+The current management of test assets is handled manually by the test
+developer. It is usual to have a set of repetitive code blocks to define the
+name, location, and other attributes of an asset, download it or raise an
+exception if a problem occurred and the download failed.
+
+Based on use cases compiled from the discussion on qemu-devel mailing-list [1]
+and discussions during avocado meetings, this blueprint describes the
+architecture of a requirements resolver aiming the extensibility and
+flexibility when handling different types of assets, like a file, a cloud
+image, a package, a Git repository, source codes or Operating System parameters.
+
+Motivation
+**********
+
+Implement a test that gathers the requirements during its execution may lead to
+a wrong interpretation of the test results if a requirement is not satisfied.
+The failure of a test because of a missing requirement does not mean the test
+itself failed. During its execution, the test has never reached the core test
+code; still, it may be considered a failing test.
+
+Fulfill all the test requirements beforehand can be an efficient way to handle
+requirements problems and can improve the trustworthiness of the test result.
+It means that if a test ran and failed, the code responsible for the failure is
+related to the core test and not with one of its requirements.
+
+Regardless of how the test defines a requirement, an architecture capable of
+identifying them is beneficial. Storing its references and delegating to the
+code responsible for handling each different type of requirement makes the
+overall architecture of Avocado and the requirement definition of a test more
+flexible.
+
+This blueprint discusses the architecture of a requirements resolver and
+defines the APIs for the modules responsible for handling the different
+requirements types.
+
+Specification
+*************
+
+Basics
+======
+
+A resolver is responsible for creating resolutions from a given representation.
```

Comment:
Maybe we could define a `resolver` and a `resolution` at some point.

Reply:
Yes, introducing the concepts will certainly help to see things more clearly.